### ISSUE TITLE
Gradle 6.0 fails to find dependencies in p2asmaven repository

### ISSUE THREAD

**Issue by liblit** [diffplug/goomph#106]
The [WALA analysis framework](https://github.com/wala/WALA) uses the [`com.diffplug.gradle.p2.asmaven`](https://plugins.gradle.org/plugin/com.diffplug.gradle.p2.asmaven) plugin to incorporate several Eclipse libraries into its build, as configured [here](https://github.com/wala/WALA/blob/3df0346b438bfbebb09aedae118cb9ad3fccd6ef/build.gradle#L31). This worked fine under Gradle 5.6.2, but under Gradle 6.0 these `p2asmaven` dependencies fail to be resolved:

```
* What went wrong:
Execution failed for task ':com.ibm.wala.ide:compileJava'.
> Could not resolve all files for configuration ':com.ibm.wala.ide:compileClasspath'.
   > Could not find any matches for eclipse-deps:org.eclipse.pde.core:+ as no versions of eclipse-deps:org.eclipse.pde.core are available.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/eclipse-deps/org.eclipse.pde.core/maven-metadata.xml
       - file:[...]/WALA/build/p2asmaven/maven/eclipse-deps/org.eclipse.pde.core/maven-metadata.xml
       - file:[...]/WALA/build/p2asmaven/maven/eclipse-deps/org.eclipse.pde.core/3.11.100.v20170517-0724/org.eclipse.pde.core-3.11.100.v20170517-0724.pom
     Required by:
         project :com.ibm.wala.ide
```
[Many more, similar “`Could not find any matches`” failures follow.]

Running a low-level system-call trace on the Gradle process shows me that both Gradle 5.6.2 and Gradle 6.0 open and read `WALA/build/p2asmaven/maven/eclipse-deps/org.eclipse.pde.core/maven-metadata.xml`, then look for (but fail to find) `WALA/build/p2asmaven/maven/eclipse-deps/org.eclipse.pde.core/3.11.100.v20170517-0724/org.eclipse.pde.core-3.11.100.v20170517-0724.pom`. After failing to find this POM file, Gradle 5.6.2 looks for and finds a Jar archive instead: `WALA/build/p2asmaven/maven/eclipse-deps/org.eclipse.pde.core/3.11.100.v20170517-0724/org.eclipse.pde.core-3.11.100.v20170517-0724.jar`. Unfortunately, Gradle 6.0 never even looks for such a file.

Is the problem here with `p2.asmaven` failing to create necessary POM files? Or should Gradle 6.0 be finding the Jar archives even without the POM files, as Gradle 5.6.2 did? Any suggestions on how I can get this `p2.asmaven`-created Maven repository back into working order for use with Gradle 6.0?

Thank you for any clues!

**Comment by nedtwigg** [comment]
You are correct, `p2.asmaven` does not generate `.pom` files.  Based on your trace, I expect that generating a `.pom` file would fix this issue.  [Here is the code that generates the `maven-metadata.xml` file](https://github.com/diffplug/goomph/blob/4ac5b997f805b020b80d1d5db092238ec8833f82/src/main/java/com/diffplug/gradle/p2/MavenRepoBuilder.java#L70-L88), and you could generate the pom file using similar code in that same file.  I'd be happy to merge a PR for this, [here is how to test locally](https://github.com/diffplug/goomph/blob/master/CONTRIBUTING.md#test-locally).

An alternative would be to dig through the gradle changelogs and/or commits, to see if they kept some kind of `permissive` or `ignorePoms` flag which would make it possible to leave the poms unimplemented.

Lowest-friction approach is probably to start with the changelogs, but the deepest fix is to generate at least a minimal pom.

Here at DiffPlug, we have transitioned off of p2 completely, and we now use the [`eclipse.mavencentral` plugin](https://diffplug.github.io/goomph/javadoc/3.18.0/com/diffplug/gradle/eclipse/MavenCentralPlugin.html) to grab the core eclipse libraries from maven central.  That doesn't help if you're using eclipse projects that don't cross-publish to maven central, though.  We'll happily help people maintain and extend the p2 support, but we won't lead any initiatives in this direction.

**Comment by liblit** [comment]
I had no idea that Eclipse had started cross-publishing to Maven central. As it turns out, nearly all Eclipse components that WALA needs are cross-published in this manner. The only exceptions are two WST JSDT libraries, which I’ve found in a different public Maven repository. So now WALA cam move off of `p2.asmaven` entirely. Ultimately our Gradle builds still need to download the Eclipse libraries, of course.  But those downloads can now be concurrent with other Gradle build activity, such
as compiling and running tests for non-Eclipse-based pieces of WALA. We also get better download progress reporting, proper Gradle caching, etc. Very nice.

Thanks for taking the time to craft an informative response, @nedtwigg. I am very happy with the alternative approach that your explanation put on my radar.

**Comment by liblit** [comment]
Gradle core developer Jendrik Johannes [expertly diagnosed the problem](https://discuss.gradle.org/t/gradle-6-0-fails-to-find-dependencies-in-p2asmaven-repository/33898/2) and recommended a fix:

> You are running into this (breaking) change:
https://docs.gradle.org/current/userguide/upgrading_version_5.html#maven_or_ivy_repositories_are_no_longer_queried_for_artifacts_without_metadata_by_default
>
> It’s a performance optiomization that now assumes that all Maven repositories follow the standard Maven layout (i.e. there always is a `maven-metadata.xml` and `pom`s)
You need to add the `artifact()` metadata sources to the `p2asmaven` repository declaration to tell Gradle that the repo is not always following these conventions.

I have not tried this myself, as I have instead moved to the `eclipse.mavencentral` plugin as [noted above](#issuecomment-558720305). But if someone else runs into this problem in the future, and cannot switch to `eclipse.mavencentral`, Jendrik’s post may help.

**Comment by nedtwigg** [comment]
Thanks for the workaround link above, it will be easy to make this the default behavior in the next release.

**Comment by nedtwigg** [comment]
Thanks for reporting and finding the workaround!  Fixed in `3.20.0`.
